state/address.go

```diff
@@ -97,39 +102,92 @@ type apiHostPortsDoc struct {
 
 // SetAPIHostPorts sets the addresses of the API server instances.
 // Each server is represented by one element in the top level slice.
-func (st *State) SetAPIHostPorts(netHostsPorts [][]network.HostPort) error {
+func (st *State) SetAPIHostPorts(newHostPorts [][]network.HostPort) error {
 	controllers, closer := st.db().GetCollection(controllersC)
 	defer closer()
-	doc := apiHostPortsDoc{
-		APIHostPorts: fromNetworkHostsPorts(netHostsPorts),
-	}
+
+	var newHostPortsForAgents [][]network.HostPort
+
 	buildTxn := func(attempt int) ([]txn.Op, error) {
-		var existingDoc apiHostPortsDoc
-		err := controllers.Find(bson.D{{"_id", apiHostPortsKey}}).One(&existingDoc)
+		var ops []txn.Op
+
+		// Retrieve the current API addresses document.
+		var existingAllAddrDoc apiHostPortsDoc
+		err := controllers.Find(bson.D{{"_id", apiHostPortsKey}}).One(&existingAllAddrDoc)
 		if err != nil {
 			return nil, err
 		}
-		op := txn.Op{
-			C:  controllersC,
-			Id: apiHostPortsKey,
-			Assert: bson.D{{
-				"txn-revno", existingDoc.TxnRevno,
-			}},
+		hostPorts := networkHostsPorts(existingAllAddrDoc.APIHostPorts)
+
+		// If the stored addresses are different from the input collection, queue an update operation.
+		if !hostsPortsEqual(newHostPorts, hostPorts) {
+			ops = append(ops, txn.Op{
+				C:  controllersC,
+				Id: apiHostPortsKey,
+				Assert: bson.D{{
+					"txn-revno", existingAllAddrDoc.TxnRevno,
+				}},
+				Update: bson.D{{
+					"$set", bson.D{{"apihostports", fromNetworkHostsPorts(newHostPorts)}},
+				}},
+			})
+		}
+
+		// Filter the collection of API addresses based on the configured management
+		// space for the controller.
+		// If there is no space configured, or if the one of the slices is filtered down
+		// to zero elements, just use the unfiltered slice for safety.
+		config, err := st.ControllerConfig()
```

Comment:
I would factor all of this out into a separate function so that it stays cohesive and doesn't end up with 10 levels of nesting.